### PR TITLE
[streaming] Async changes for resourcemanager part

### PR DESCRIPTION
**Pull request by datayjz**
## Why are these changes needed?
We did some changes on 'graph', 'resource' and 'scheduler' in our internal code base(streaming-runtime) recently for better structure.
So this PR is mainly to sync 'resource' part's changes.

The major changes are:
1. Removed ```Slot```.
2. Reimplemented ```AbstractId```.
3. Optimized the data structure of ```Resources``` and ````Container```. 
4.  Redesigned and implemented ```ResourceManager```, making the resource allocation process more convenient.
5. Optimize ut.
## Related issue number
#6184 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/ResourceAssignmentView.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.ray.streaming.runtime.master.resourcemanager;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.ray.streaming.runtime.core.resource.ContainerID;
+
+public class ResourceAssignmentView extends HashMap<ContainerID, List<Integer>> {
```
Comment (BalaBalaYi) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/ResourceAssignmentView.java:8]: Need some comment for this class.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/ResourceManagerImpl.java [hunk 2]
```diff
@@ -41,160 +44,153 @@
...
+      registerNewContainers(addNodes.stream().map(latestNodeInfos::get).collect(Collectors.toList()));
     }
-    //Clear deleted nodes
-    if (!deleteContainers.isEmpty()) {
-      for (Container container : deleteContainers) {
-        unregisterContainer(container);
-      }
+  }
+
+
```
Comment (BalaBalaYi) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/ResourceManagerImpl.java:137]: Remove blank.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/strategy/ResourceAssignStrategy.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package org.ray.streaming.runtime.master.resourcemanager.strategy;
+
+import java.util.List;
+import org.ray.streaming.runtime.core.graph.executiongraph.ExecutionGraph;
+import org.ray.streaming.runtime.core.resource.Container;
+import org.ray.streaming.runtime.master.resourcemanager.ResourceAssignmentView;
+
+/**
+ * The ResourceAssignStrategy responsible assign {@link Container} to
+ * {@link org.ray.streaming.runtime.core.graph.executiongraph.ExecutionVertex}.
```
Comment (BalaBalaYi) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/strategy/ResourceAssignStrategy.java:10]: Move the full qualified class name into import.
e.g.
import xxx.xxx.A
{@link A}

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/strategy/ResourceAssignStrategy.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  /**
+   * Assign {@link Container} for {@link org.ray.streaming.runtime.core.graph.executiongraph.ExecutionVertex}
+   *
+   * @param containers registered container
+   * @param executionGraph execution graph
+   * @return allocating view
+   */
+  ResourceAssignmentView assignResource(List<Container> containers, ExecutionGraph executionGraph);
+
+
```
Comment (BalaBalaYi) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/strategy/ResourceAssignStrategy.java:23]: Remove blank.

File: streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/util/Mockitools.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
+package org.ray.streaming.runtime.util;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+
```
Comment (BalaBalaYi) [inline streaming/java/streaming-runtime/src/test/java/org/ray/streaming/runtime/util/Mockitools.java:9]: Remove blank.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/core/graph/executiongraph/ExecutionVertex.java [hunk 6]
```diff
@@ -182,14 +182,19 @@ public void setSlotIfNotExist(Slot slot) {
     return resourceMap;
   }
 
+  @Override
+  public boolean equals(Object obj) {
+    return this.id == ((ExecutionVertex)obj).getId();
```
Comment (chaokunyang) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/core/graph/executiongraph/ExecutionVertex.java:187]: should check `obj` type first

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/ResourceManager.java [hunk 7]
```diff
@@ -1,53 +1,19 @@
...
 
   /**
-   * Get all registered container as a list.
+   * Get registered containers
    *
-   * @return A list of containers.
+   * @return the registered container list
    */
-  List<Container> getRegisteredContainers();
+  ImmutableList<Container> getRegisteredContainers();
```
Comment (chaokunyang) [inline streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/ResourceManager.java:17]: I found many interface methods return `ImmutableList/ImmutableMap` . It's necessary to emphasize it's immutable or just changed return type to `List/Map`

Comment (chaokunyang) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/ResourceManager.java:17]: It's more a taste

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/strategy/ResourceAssignStrategy.java (hunk 3 above)
Comment (datayjz) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/strategy/ResourceAssignStrategy.java]: Full name more better?

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/master/ResourceConfig.java [hunk 8]
```diff
@@ -70,7 +70,7 @@
    */
   @DefaultValue(value = "500")
   @Key(MAX_ACTOR_NUM_PER_CONTAINER)
-  int maxActorNumPerContainer();
+  int customActorNumPerContainer();
```
[the remark this answers is not in the document]
Comment (datayjz) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/config/master/ResourceConfig.java]: yes,  already updated.

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/ResourceManager.java (hunk 7 above)
Comment (datayjz) [reply streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/master/resourcemanager/ResourceManager.java]: Yes, comments have been added to related methods.

**Review by chaokunyang — APPROVED** [review]
LGTM, thanks

**Review by raulchen — APPROVED** [review]

**Pull request merged**